doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,268 @@
+period of time between when the server application responds with a Trailers and
+when that Trailers is actually sent; the Call is only truly complete when the
+Trailers is sent. Similarly, on the client there is a period between the gRPC
+implementation receiving the Trailers and when the application receives the
+Trailers; the Call is only truly complete when the Trailers is received by the
+application.
+
+Calls may terminate early by being "cancelled." Implementations must allow
+clients to cancel Calls, but cancellations may occur in other ways like I/O
+failures. A cancellation appears as a Trailers with a Status Code of CANCELLED
+to clients and is a special state on servers. Cancellation is an abrupt killing
+of the Call; inbound and outbound buffered data should be cleared. Cancellation
+trumps graceful completion; if the client gRPC implementation received the
+Trailers before the cancellation, yet the client application has not received
+the Trailers, then cancellation generally should win. No auxiliary information
+is included in cancellation signals between the client and server. Server
+implementations may fail a Call and respond with Trailers while claiming to the
+server application that the Call was cancelled.
+
+The two independent streams are each unidirectional and do not provide any
+information in the reverse direction other than Flow Control. Flow Control
+is a signal from the receiver to the sender to temporarily pause sending
+additional messages to avoid excessive buffering. Flow Control only applies to
+messages, but since streams are in-order Half Close or Trailers may be delayed
+waiting for message Flow Control in the same stream. No message receipt
+acknowledgement information is provided. However applications may use messages
+for such signals, as a response naturally acknowledges its request. Note in
+particular that there is no provision for the server application to not know
+whether the client application received a unary Call's response or a streaming
+Call's Status.
+
+Unary Calls may be optimized to be half-duplex and treat each stream as a single
+communication unit. That is, on the client a unary Call may be delayed from
+being sent until the Half Close is ready to be sent and on the server the
+response may be delayed until the Trailers is ready.
+
+Unary Calls that terminate with a Status Code of OK must contain a response
+message. Unary Calls that terminate with a Status Code other than OK do not need
+a response message, and at the implementation's discretion the response message
+may be discarded if present.
+
+## Status
+
+A Status contains a "code" and a "description". The Status Description is a
+human-readable Unicode string for developer debugging. The Status Code is a
+value from a pre-defined list of such codes. While Status Code is best
+communicated to users by its name, it commonly is treated as an integer
+internally, and so each code has a numeric value.
```

Comment:
I think we should have something here about: even though it is treated as an integer, any value outside this range is coerced to UNKNOWN (2) before reaching the application layer.